user_sync/connector/directory_ldap.py

```diff
@@ -132,35 +144,52 @@ def load_users_and_groups(self, groups, extended_attributes, all_users):
+        two_steps_lookup = self.two_steps_lookup
+
+        # save all the users to memory for faster 2-steps lookup process
+        if all_users or two_steps_lookup:
+            all_users_records = dict(self.iter_users(all_users_filter, extended_attributes))
 
         # for each group that's required, do one search for the users of that group
         for group in groups:
-            group_dn = self.find_ldap_group_dn(group)
-            if not group_dn:
-                self.logger.warning("No group found for: %s", group)
-                continue
-            group_member_subfilter = self.format_ldap_query_string(group_member_filter_format, group_dn=group_dn)
-            if not group_member_subfilter.startswith('('):
-                group_member_subfilter = six.text_type('(') + group_member_subfilter + six.text_type(')')
-            user_subfilter = all_users_filter
-            if not user_subfilter.startswith('('):
-                user_subfilter = six.text_type('(') + user_subfilter + six.text_type(')')
-            group_user_filter = six.text_type('(&') + group_member_subfilter + user_subfilter + six.text_type(')')
             group_users = 0
-            try:
-                for user_dn, user in self.iter_users(group_user_filter, extended_attributes):
-                    user['groups'].append(group)
-                    group_users += 1
-                self.logger.debug('Count of users in group "%s": %d', group, group_users)
-            except Exception as e:
-                raise AssertionException('Unexpected LDAP failure reading group members: %s' % e)
+            if not two_steps_lookup:
+                group_dn = self.find_group_dn(group)
+                if not group_dn:
+                    self.logger.warning("No group found for: %s", group)
+                    continue
+                group_member_subfilter = self.format_ldap_query_string(group_member_filter_format, group_dn=group_dn)
+                if not group_member_subfilter.startswith('('):
+                    group_member_subfilter = six.text_type('(') + group_member_subfilter + six.text_type(')')
+                user_subfilter = all_users_filter
+                if not user_subfilter.startswith('('):
+                    user_subfilter = six.text_type('(') + user_subfilter + six.text_type(')')
+                group_user_filter = six.text_type('(&') + group_member_subfilter + user_subfilter + six.text_type(')')
+                try:
+                    for user_dn, user in self.iter_users(group_user_filter, extended_attributes):
+                        user['groups'].append(group)
+                        group_users += 1
+                except Exception as e:
+                    raise AssertionException('Unexpected LDAP failure reading group members: %s' % e)
+            else:
+                try:
+                    members = self.find_group_member_dn(group, two_steps_group_member_attribute_name)
```

Comment:
this should be called `iter_group_member_dns`and don't assign it to a list and then iterate the list, instead do:
```python
for member in self.iter_group_member_dns(group, group_member_attribute_name)
```